### PR TITLE
ENH: add from_dataframe and DataFrame

### PR DESCRIPTION
**Pull request by bashtage**
- [X] xref #227

### CODE REVIEW

File: pandas-stubs/core/interchange/dataframe_protocol.pyi [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    @property
+    @abstractmethod
+    def bufsize(self) -> int: ...
+    @property
+    @abstractmethod
+    def ptr(self) -> int: ...
+    @abstractmethod
+    def __dlpack__(self): ...
+    @abstractmethod
+    def __dlpack_device__(self) -> tuple[DlpackDeviceType, int | None]: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/interchange/dataframe_protocol.pyi:60]: Looking at https://data-apis.org/dataframe-protocol/latest/API.html, I don't think the second element of the tuple can be `None`

File: pandas-stubs/core/interchange/dataframe_protocol.pyi [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+class Column(ABC, metaclass=abc.ABCMeta):
+    @property
+    @abstractmethod
+    def size(self) -> int: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/interchange/dataframe_protocol.pyi:65]: Looking at https://data-apis.org/dataframe-protocol/latest/API.html, returns `Optional[int]`

File: pandas-stubs/core/interchange/dataframe_protocol.pyi [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+class ColumnBuffers(TypedDict):
+    data: tuple[Buffer, Any]
+    validity: tuple[Buffer, Any] | None
+    offsets: tuple[Buffer, Any] | None
+
+class CategoricalDescription(TypedDict):
+    is_ordered: bool
+    is_dictionary: bool
+    categories: Column | None
```
Comment (Dr-Irv) [inline pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: Looking at https://data-apis.org/dataframe-protocol/latest/API.html, `categories` should be `dict | None`

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Unclear to me whether we should add tests for this?  Other comments are related to checking the API definition at https://data-apis.org/dataframe-protocol/latest/API.html, which may reveal that we have issues in the pandas source in how this is typed.

File: pandas-stubs/core/interchange/dataframe_protocol.pyi (hunk 3 above)
Comment (bashtage) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: All of the types here are straight from pandas main.

Comment (Dr-Irv) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: They may be wrong based on the docs, and are probably not tested within pandas itself, so can we fix here and do a PR there?

Comment (bashtage) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: Here https://github.com/pandas-dev/pandas/blob/396ad412cf4af59af199101d9b5e1e74f45c698a/pandas/core/interchange/dataframe_protocol.py#L273 it indicates that it should be a `Column`.

File: pandas-stubs/core/interchange/dataframe_protocol.pyi (hunk 2 above)
Comment (bashtage) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:65]: I think the spec type might also be wrong here.  The text indicates that the size is required, either the full DF size or the chunk size if a single chunk.

File: pandas-stubs/core/interchange/dataframe_protocol.pyi (hunk 1 above)
Comment (bashtage) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:60]: Here

https://github.com/pandas-dev/pandas/blob/a046792de912146582828de1baf3008750d93bd0/pandas/core/interchange/buffer.py#L60

it is `None`

File: pandas-stubs/core/interchange/dataframe_protocol.pyi (hunk 3 above)
Comment (Dr-Irv) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: I used this to verify the API:  https://data-apis.org/dataframe-protocol/latest/API.html

I think we may have errors in the pandas implementation _if_ that is the "official" standard.

Comment (bashtage) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: I suspect the code in pandas is the de facto even if it differs from the docs.

Comment (bashtage) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: I would also add that I think pandas-stubs should match pandas, and if there is an issue, pandas should change first, then pandas-stubs keep up.

Comment (Dr-Irv) [reply pandas-stubs/core/interchange/dataframe_protocol.pyi:48]: Did a little research, and it appears the docs I looked at are not up to date.  The interface changed in https://github.com/data-apis/dataframe-api/pull/69

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

**Pull request merged**